<?php declare(strict_types = 1);

namespace App\Services\DataLoader\Client;

use Exception;
use PHPUnit\Framework\TestCase;

/**
 * @internal
 * @coversDefaultClass \App\Services\DataLoader\Client\QueryIteratorImpl
 */
class QueryIteratorImplTest extends TestCase {
    /**
     * @covers ::chunkLoaded
     * @covers ::onBeforeChunk
     */
    public function testChunkLoaded(): void {
        $exception = new Exception(__METHOD__);
        $iterator  = new class() extends QueryIteratorImpl {
            /** @noinspection PhpMissingParentConstructorInspection */
            public function __construct() {
                // empty
            }

            /**
             * @inheritDoc
             */
            protected function getQueryParams(): array {
                return [];
            }

            /**
             * @inheritDoc
             */
            public function chunkLoaded(array $items): void {
                parent::chunkLoaded($items);
            }
        };

        $this->expectExceptionObject($exception);

        $iterator->onBeforeChunk(static function () use ($exception): void {
            throw $exception;
        });
        $iterator->chunkLoaded([]);
    }

    /**
     * @covers ::chunkProcessed
     * @covers ::onAfterChunk
     */
    public function testChunkProcessed(): void {
        $exception = new Exception(__METHOD__);
        $iterator  = new class() extends QueryIteratorImpl {
            /** @noinspection PhpMissingParentConstructorInspection */
            public function __construct() {
                // empty
            }

            /**
             * @inheritDoc
             */
            protected function getQueryParams(): array {
                return [];
            }

            /**
             * @inheritDoc
             */
            public function chunkProcessed(array $items): bool {
                return parent::chunkProcessed($items); // TODO: Change the autogenerated stub
            }
        };

        $this->expectExceptionObject($exception);

        $iterator->onAfterChunk(static function () use ($exception): void {
            throw $exception;
        });
        $iterator->chunkProcessed([]);
    }
}
